README.md
classList.js is a cross-browser JavaScript shim that fully implements `element.classList`. Refer to [the MDN page on `element.classList`][1] for more information.

This works in every browser except IE 7 or earlier.

The current version is hosted at public CDNs, allowing you to use this already small file at nearly zero size overhead. Use one of these URLs:

  - [//cdnjs.cloudflare.com/ajax/libs/classlist/2014.07.23/classList.min.js](//cdnjs.cloudflare.com/ajax/libs/classlist/2014.01.31/classList.min.js)
  - [//cdn.jsdelivr.net/classlist/2014.07.23/classList.min.js](//cdn.jsdelivr.net/classlist/2014.01.31/classList.min.js)

![Tracking image](https://in.getclicky.com/212712ns.gif)

  [1]: https://developer.mozilla.org/en/DOM/element.classList "MDN / DOM / element.classList"


CHANGELOG.md
# Changelog

## Dev

* Rewrite history for [yola/classlist-polyfill][], branching and rebasing
  from [eligrey/classList][].

[yola/classlist-polyfill]: https://github.com/yola/classlist-polyfill
[eligrey/classList]: https://github.com/eligrey/classList.js


## 1.0.3
* Add support for missing SVGElement.classList in IE


## 1.0.2
* Fix issue with `self` not being defined in CommonJS


tests/add.js
QUnit.module("classList.toggle");

QUnit.test("Adds a class", function(assert) {
	var p = document.createElement("p");
	var g = document.createElementNS("http://www.w3.org/2000/svg","svg").appendChild( document.createElement("g") );
	var gList = g.classList;
	var pList = p.classList;

	pList.add("c1");
	assert.ok(pList.contains("c1"), "Adds a class that is not present in HTML element");
	assert.strictEqual(p.className, "c1");

	gList.add("c1");
	assert.ok(gList.contains("c1"), "Adds a class that is not present in SVG element");
	assert.strictEqual(g.className, "c1");
});

QUnit.test("Adds multiple classes", function(assert) {
	var p = document.createElement("p");
	var g = document.createElementNS("http://www.w3.org/2000/svg","svg").appendChild( document.createElement("g") );
	var gList = g.classList;
	var pList = p.classList;

	pList.add("c1", "c2");
	assert.ok(pList.contains("c1") && pList.contains("c2"), "Adds classes that are not present in HTML element");
	assert.strictEqual(p.className, "c1 c2");

	gList.add("c1", "c2");
	assert.ok(gList.contains("c1") && gList.contains("c2"), "Adds classes that are not present in SVG element");
	assert.strictEqual(g.className, "c1 c2");
});


tests/tests.js
QUnit.module("classList.remove");

QUnit.test("Removes duplicated instances of class", function(assert) {
	var el = document.createElement("p");
	el.className = "ho ho ho"
	var cList = el[classListProp];

	cList.remove("ho");
	assert.ok(!cList.contains("ho"), "Should remove all instances of 'ho'");
	assert.strictEqual(el.className, "")
});

QUnit.module("classList.toggle");

QUnit.test("Adds a class", function(assert) {
	var cList = document.createElement("p")[classListProp];

	cList.toggle("c1");
	assert.ok(cList.contains("c1"), "Adds a class that is not present");

	assert.strictEqual(
		cList.toggle("c2"),
		true,
		"Returns true when class is added"
	);
});

QUnit.test("Removes a class", function(assert) {
	var cList = document.createElement("p")[classListProp];

	cList.add("c1");
	cList.toggle("c1");
	assert.ok(!cList.contains("c1"), "Removes a class that is present");

	cList.add("c2");
	assert.strictEqual(
		cList.toggle("c2"),
		false,
		"Return false when class is removed"
	);
});

QUnit.test("Adds class with second argument", function(assert) {
	var cList = document.createElement("p")[classListProp];

	cList.toggle("c1", true);
	assert.ok(cList.contains("c1"), "Adds a class");

	assert.strictEqual(
		cList.toggle("c2", true),
		true,
		"Returns true when class is added"
	);

	cList.add("c3");
	cList.toggle("c3", true);
	assert.ok(
		cList.contains("c3"),
		"Does not remove a class that is already present"
	);

	cList.add("c4");
	assert.strictEqual(
		cList.toggle("c4", true),
		true,
		"Returns true when class is already present"
	);
});

QUnit.test("Removes class with second argument", function(assert) {
	var cList = document.createElement("p")[classListProp];

	cList.add("c1");
	cList.toggle("c1", false);
	assert.ok(!cList.contains("c1"), "Removes a class");

	assert.strictEqual(
		cList.toggle("c2", false),
		false,
		"Returns false when class is removed"
	);

	cList.toggle("c3", false);
	assert.ok(
		!cList.contains("c3"),
		"Does not add a class that is not present"
	);

	assert.strictEqual(
		cList.toggle("c4", false),
		false,
		"Returns false when class was not present"
	);
});


tests/remove.js
QUnit.module("classList.remove");

QUnit.test("Removes a class", function(assert) {
  var p = document.createElement("p");
  var g = document.createElementNS("http://www.w3.org/2000/svg","svg").appendChild( document.createElement("g") );
  var gList = g.classList;
  var pList = p.classList;
  g.className = "c1 c2";
  p.className = "c1 c2";

  pList.remove("c1");
  assert.ok(!pList.contains("c1"), "Removes a class that is present in HTML element");
  assert.strictEqual(p.className, "c2");

  gList.remove("c1");
  assert.ok(!gList.contains("c1"), "Removes a class that is present in SVG element");
  assert.strictEqual(g.className, "c2");
});

QUnit.test("Removes multiple classes", function(assert) {
  var p = document.createElement("p");
  var g = document.createElementNS("http://www.w3.org/2000/svg","svg").appendChild( document.createElement("g") );
  var gList = g.classList;
  var pList = p.classList;
  g.className = "c1 c2 c3";
  p.className = "c1 c2 c3";

  pList.remove("c1", "c3");
  assert.ok(!pList.contains("c1") && !pList.contains("c3"), "Removes mutliple classes that are present in HTML element");
  assert.strictEqual(p.className, "c2");

  gList.remove("c1", "c3");
  assert.ok(!gList.contains("c1") && !gList.contains("c3"), "Removes mutliple classes that are present in SVG element");
  assert.strictEqual(g.className, "c2");
});

QUnit.test("Removes duplicated instances of class", function(assert) {
  var p = document.createElement("p");
  var g = document.createElementNS("http://www.w3.org/2000/svg","svg").appendChild( document.createElement("g") );
  var gList = g.classList;
  var pList = p.classList;
  g.className = "c1 c1 c1";
  p.className = "c1 c1 c1";

	pList.remove("c1");
	assert.ok(!pList.contains("c1"), "Removes all duplicated instances of classes that are present in HTML element");
	assert.strictEqual(p.className, "");

  gList.remove("c1");
  assert.ok(!gList.contains("c1"), "Removes all duplicated instances of classes that are present in SVG element");
  assert.strictEqual(g.className, "");
});


classList.js
/*
 * classList.js: Cross-browser full element.classList implementation.
 * 1.1.20160404
 *
 * By Eli Grey, http://eligrey.com
 * License: Dedicated to the public domain.
 *   See https://github.com/eligrey/classList.js/blob/master/LICENSE.md
 */

/*global self, document, DOMException */

/*! @source http://purl.eligrey.com/github/classList.js/blob/master/classList.js */

if ("document" in self) {
    (function() {
        "use strict";

        var setPropOnDOM = true;

        // Detect if we can define properties on DOM Objects
        try {
            Object.defineProperty(document.createElement("_"), "classList", {
                "get": function() {
                }
            });
        } catch (e) { //Old Safari versions will break at this point
            setPropOnDOM = false;
        }

// Full polyfill for browsers with no classList support
// Including IE < Edge missing SVGElement.classList
        if (setPropOnDOM && (!("classList" in document.createElement("_")) || document.createElementNS && !("classList" in document.createElementNS("http://www.w3.org/2000/svg", "g")))) {

            (function(view) {

                "use strict";

                if (!('Element' in view)) return;

                var
                    classListProp = "classList"
                    , protoProp = "prototype"
                    , elemCtrProto = view.Element[protoProp]
                    , objCtr = Object
                    , strTrim = String[protoProp].trim || function() {
                            return this.replace(/^\s+|\s+$/g, "");
                        }
                    , arrIndexOf = Array[protoProp].indexOf || function(item) {
                            var
                                i = 0
                                , len = this.length
                                ;
                            for (; i < len; i++) {
                                if (i in this && this[i] === item) {
                                    return i;
                                }
                            }
                            return -1;
                        }
                // Vendors: please allow content code to instantiate DOMExceptions
                    , DOMEx = function(type, message) {
                        this.name = type;
                        this.code = DOMException[type];
                        this.message = message;
                    }
                    , checkTokenAndGetIndex = function(classList, token) {
                        if (token === "") {
                            throw new DOMEx(
                                "SYNTAX_ERR"
                                , "An invalid or illegal string was specified"
                            );
                        }
                        if (/\s/.test(token)) {
                            throw new DOMEx(
                                "INVALID_CHARACTER_ERR"
                                , "String contains an invalid character"
                            );
                        }
                        return arrIndexOf.call(classList, token);
                    }
                    , ClassList = function(elem) {
                        var
                            trimmedClasses = strTrim.call(elem.getAttribute("class") || "")
                            , classes = trimmedClasses ? trimmedClasses.split(/\s+/) : []
                            , i = 0
                            , len = classes.length
                            ;
                        for (; i < len; i++) {
                            this.push(classes[i]);
                        }
                        this._updateClassName = function() {
                            elem.setAttribute("class", this.toString());
                        };
                    }
                    , classListProto = ClassList[protoProp] = []
                    , classListGetter = function() {
                        return new ClassList(this);
                    }
                    ;
// Most DOMException implementations don't allow calling DOMException's toString()
// on non-DOMExceptions. Error's toString() is sufficient here.
                DOMEx[protoProp] = Error[protoProp];
                classListProto.item = function(i) {
                    return this[i] || null;
                };
                classListProto.contains = function(token) {
                    token += "";
                    return checkTokenAndGetIndex(this, token) !== -1;
                };
                classListProto.add = function() {
                    var
                        tokens = arguments
                        , i = 0
                        , l = tokens.length
                        , token
                        , updated = false
                        ;
                    do {
                        token = tokens[i] + "";
                        if (checkTokenAndGetIndex(this, token) === -1) {
                            this.push(token);
                            updated = true;
                        }
                    }
                    while (++i < l);

                    if (updated) {
                        this._updateClassName();
                    }
                };
                classListProto.remove = function() {
                    var
                        tokens = arguments
                        , i = 0
                        , l = tokens.length
                        , token
                        , updated = false
                        , index
                        ;
                    do {
                        token = tokens[i] + "";
                        index = checkTokenAndGetIndex(this, token);
                        while (index !== -1) {
                            this.splice(index, 1);
                            updated = true;
                            index = checkTokenAndGetIndex(this, token);
                        }
                    }
                    while (++i < l);

                    if (updated) {
                        this._updateClassName();
                    }
                };
                classListProto.toggle = function(token, force) {
                    token += "";

                    var
                        result = this.contains(token)
                        , method = result ?
                        force !== true && "remove"
                            :
                        force !== false && "add"
                        ;

                    if (method) {
                        this[method](token);
                    }

                    if (force === true || force === false) {
                        return force;
                    } else {
                        return !result;
                    }
                };
                classListProto.toString = function() {
                    return this.join(" ");
                };

                if (objCtr.defineProperty) {
                    var classListPropDesc = {
                        get: classListGetter
                        , enumerable: true
                        , configurable: true
                    };
                    try {
                        objCtr.defineProperty(elemCtrProto, classListProp, classListPropDesc);
                    } catch (ex) { // IE 8 doesn't support enumerable:true
                        if (ex.number === -0x7FF5EC54) {
                            classListPropDesc.enumerable = false;
                            objCtr.defineProperty(elemCtrProto, classListProp, classListPropDesc);
                        }
                    }
                } else if (objCtr[protoProp].__defineGetter__) {
                    elemCtrProto.__defineGetter__(classListProp, classListGetter);
                }

            }(self));

        } else {
// There is full or partial native classList support, so just check if we need
// to normalize the add/remove and toggle APIs.

            (function() {
                "use strict";

                var testElement = document.createElement("_");

                if (!!("classList" in testElement)) {

                    testElement.classList.add("c1", "c2");

                    // Polyfill for IE 10/11 and Firefox <26, where classList.add and
                    // classList.remove exist but support only one argument at a time.
                    if (!testElement.classList.contains("c2")) {
                        var createMethod = function(method) {
                            var original = DOMTokenList.prototype[method];

                            DOMTokenList.prototype[method] = function(token) {
                                var i, len = arguments.length;

                                for (i = 0; i < len; i++) {
                                    token = arguments[i];
                                    original.call(this, token);
                                }
                            };
                        };
                        createMethod('add');
                        createMethod('remove');
                    }

                    testElement.classList.toggle("c3", false);

                    // Polyfill for IE 10 and Firefox <24, where classList.toggle does not
                    // support the second argument.
                    if (testElement.classList.contains("c3")) {
                        var _toggle = DOMTokenList.prototype.toggle;

                        DOMTokenList.prototype.toggle = function(token, force) {
                            if (1 in arguments && !this.contains(token) === !force) {
                                return force;
                            } else {
                                return _toggle.call(this, token);
                            }
                        };

                    }

                    testElement = null;
                }
            }());
        }
    })();
}



svgclasslist.js
/*
 * classList.js: Cross-browser full element.classList implementation.
 * 2012-11-15
 *
 * By Eli Grey, http://eligrey.com
 * Public Domain.
 * NO WARRANTY EXPRESSED OR IMPLIED. USE AT YOUR OWN RISK.
 */

// Modified to add classList support to SVG elements.
// Required by (older) WebKit browsers, such as Safari 5.1 and Android browser.
// Not required by Chrome, Firefox, Opera.

/*global self, document, DOMException */

/*! @source http://purl.eligrey.com/github/classList.js/blob/master/classList.js*/

if (typeof document !== "undefined" &&
	document.getElementsByTagName("svg").length > 0 &&
	!("classList" in document.getElementsByTagName("svg")[0])) {

(function (view) {

"use strict";

if (!('SVGElement' in view)) return;

var
	  classListProp = "classList"
	, protoProp = "prototype"
	, elemCtrProto = (view.SVGElement)[protoProp]
	, objCtr = Object
	, strTrim = String[protoProp].trim || function () {
		return this.replace(/^\s+|\s+$/g, "");
	}
	, arrIndexOf = Array[protoProp].indexOf || function (item) {
		var
			  i = 0
			, len = this.length
		;
		for (; i < len; i++) {
			if (i in this && this[i] === item) {
				return i;
			}
		}
		return -1;
	}
	// Vendors: please allow content code to instantiate DOMExceptions
	, DOMEx = function (type, message) {
		this.name = type;
		this.code = DOMException[type];
		this.message = message;
	}
	, checkTokenAndGetIndex = function (classList, token) {
		if (token === "") {
			throw new DOMEx(
				  "SYNTAX_ERR"
				, "An invalid or illegal string was specified"
			);
		}
		if (/\s/.test(token)) {
			throw new DOMEx(
				  "INVALID_CHARACTER_ERR"
				, "String contains an invalid character"
			);
		}
		return arrIndexOf.call(classList, token);
	}
	, ClassList = function (elem) {
		var
			  trimmedClasses = strTrim.call(elem.getAttribute('class'))
			, classes = trimmedClasses ? trimmedClasses.split(/\s+/) : []
			, i = 0
			, len = classes.length
		;
		for (; i < len; i++) {
			this.push(classes[i]);
		}
		this._updateClassName = function () {
			elem.setAttribute('class', this.toString());
		};
	}
	, classListProto = ClassList[protoProp] = []
	, classListGetter = function () {
		return new ClassList(this);
	}
;
// Most DOMException implementations don't allow calling DOMException's toString()
// on non-DOMExceptions. Error's toString() is sufficient here.
DOMEx[protoProp] = Error[protoProp];
classListProto.item = function (i) {
	return this[i] || null;
};
classListProto.contains = function (token) {
	token += "";
	return checkTokenAndGetIndex(this, token) !== -1;
};
classListProto.add = function () {
	var
		  tokens = arguments
		, i = 0
		, l = tokens.length
		, token
		, updated = false
	;
	do {
		token = tokens[i] + "";
		if (checkTokenAndGetIndex(this, token) === -1) {
			this.push(token);
			updated = true;
		}
	}
	while (++i < l);

	if (updated) {
		this._updateClassName();
	}
};
classListProto.remove = function () {
	var
		  tokens = arguments
		, i = 0
		, l = tokens.length
		, token
		, updated = false
	;
	do {
		token = tokens[i] + "";
		var index = checkTokenAndGetIndex(this, token);
		if (index !== -1) {
			this.splice(index, 1);
			updated = true;
		}
	}
	while (++i < l);

	if (updated) {
		this._updateClassName();
	}
};
classListProto.toggle = function (token, forse) {
	token += "";

	var
		  result = this.contains(token)
		, method = result ?
			forse !== true && "remove"
		:
			forse !== false && "add"
	;

	if (method) {
		this[method](token);
	}

	return !result;
};
classListProto.toString = function () {
	return this.join(" ");
};

if (objCtr.defineProperty) {
	var classListPropDesc = {
		  get: classListGetter
		, enumerable: true
		, configurable: true
	};
	try {
		objCtr.defineProperty(elemCtrProto, classListProp, classListPropDesc);
	} catch (ex) { // IE 8 doesn't support enumerable:true
		if (ex.number === -0x7FF5EC54) {
			classListPropDesc.enumerable = false;
			objCtr.defineProperty(elemCtrProto, classListProp, classListPropDesc);
		}
	}
} else if (objCtr[protoProp].__defineGetter__) {
	elemCtrProto.__defineGetter__(classListProp, classListGetter);
}

}(self));

}


src/index.js
/*
 * classList.js: Cross-browser full element.classList implementation.
 * 1.1.20150312
 *
 * By Eli Grey, http://eligrey.com
 * License: Dedicated to the public domain.
 *   See https://github.com/eligrey/classList.js/blob/master/LICENSE.md
 */

/*global self, document, DOMException */

/*! @source http://purl.eligrey.com/github/classList.js/blob/master/classList.js */

if ("document" in window.self) {

// Full polyfill for browsers with no classList support
// Including IE < Edge missing SVGElement.classList
if (!("classList" in document.createElement("_")) 
	|| document.createElementNS && !("classList" in document.createElementNS("http://www.w3.org/2000/svg","g"))) {

(function (view) {

"use strict";

if (!('Element' in view)) return;

var
	  classListProp = "classList"
	, protoProp = "prototype"
	, elemCtrProto = view.Element[protoProp]
	, objCtr = Object
	, strTrim = String[protoProp].trim || function () {
		return this.replace(/^\s+|\s+$/g, "");
	}
	, arrIndexOf = Array[protoProp].indexOf || function (item) {
		var
			  i = 0
			, len = this.length
		;
		for (; i < len; i++) {
			if (i in this && this[i] === item) {
				return i;
			}
		}
		return -1;
	}
	// Vendors: please allow content code to instantiate DOMExceptions
	, DOMEx = function (type, message) {
		this.name = type;
		this.code = DOMException[type];
		this.message = message;
	}
	, checkTokenAndGetIndex = function (classList, token) {
		if (token === "") {
			throw new DOMEx(
				  "SYNTAX_ERR"
				, "An invalid or illegal string was specified"
			);
		}
		if (/\s/.test(token)) {
			throw new DOMEx(
				  "INVALID_CHARACTER_ERR"
				, "String contains an invalid character"
			);
		}
		return arrIndexOf.call(classList, token);
	}
	, ClassList = function (elem) {
		var
			  trimmedClasses = strTrim.call(elem.getAttribute("class") || "")
			, classes = trimmedClasses ? trimmedClasses.split(/\s+/) : []
			, i = 0
			, len = classes.length
		;
		for (; i < len; i++) {
			this.push(classes[i]);
		}
		this._updateClassName = function () {
			elem.setAttribute("class", this.toString());
		};
	}
	, classListProto = ClassList[protoProp] = []
	, classListGetter = function () {
		return new ClassList(this);
	}
;
// Most DOMException implementations don't allow calling DOMException's toString()
// on non-DOMExceptions. Error's toString() is sufficient here.
DOMEx[protoProp] = Error[protoProp];
classListProto.item = function (i) {
	return this[i] || null;
};
classListProto.contains = function (token) {
	token += "";
	return checkTokenAndGetIndex(this, token) !== -1;
};
classListProto.add = function () {
	var
		  tokens = arguments
		, i = 0
		, l = tokens.length
		, token
		, updated = false
	;
	do {
		token = tokens[i] + "";
		if (checkTokenAndGetIndex(this, token) === -1) {
			this.push(token);
			updated = true;
		}
	}
	while (++i < l);

	if (updated) {
		this._updateClassName();
	}
};
classListProto.remove = function () {
	var
		  tokens = arguments
		, i = 0
		, l = tokens.length
		, token
		, updated = false
		, index
	;
	do {
		token = tokens[i] + "";
		index = checkTokenAndGetIndex(this, token);
		while (index !== -1) {
			this.splice(index, 1);
			updated = true;
			index = checkTokenAndGetIndex(this, token);
		}
	}
	while (++i < l);

	if (updated) {
		this._updateClassName();
	}
};
classListProto.toggle = function (token, force) {
	token += "";

	var
		  result = this.contains(token)
		, method = result ?
			force !== true && "remove"
		:
			force !== false && "add"
	;

	if (method) {
		this[method](token);
	}

	if (force === true || force === false) {
		return force;
	} else {
		return !result;
	}
};
classListProto.toString = function () {
	return this.join(" ");
};

if (objCtr.defineProperty) {
	var classListPropDesc = {
		  get: classListGetter
		, enumerable: true
		, configurable: true
	};
	try {
		objCtr.defineProperty(elemCtrProto, classListProp, classListPropDesc);
	} catch (ex) { // IE 8 doesn't support enumerable:true
		if (ex.number === -0x7FF5EC54) {
			classListPropDesc.enumerable = false;
			objCtr.defineProperty(elemCtrProto, classListProp, classListPropDesc);
		}
	}
} else if (objCtr[protoProp].__defineGetter__) {
	elemCtrProto.__defineGetter__(classListProp, classListGetter);
}

}(window.self));

} else {
// There is full or partial native classList support, so just check if we need
// to normalize the add/remove and toggle APIs.

(function () {
	"use strict";

	var testElement = document.createElement("_");

	testElement.classList.add("c1", "c2");

	// Polyfill for IE 10/11 and Firefox <26, where classList.add and
	// classList.remove exist but support only one argument at a time.
	if (!testElement.classList.contains("c2")) {
		var createMethod = function(method) {
			var original = DOMTokenList.prototype[method];

			DOMTokenList.prototype[method] = function(token) {
				var i, len = arguments.length;

				for (i = 0; i < len; i++) {
					token = arguments[i];
					original.call(this, token);
				}
			};
		};
		createMethod('add');
		createMethod('remove');
	}

	testElement.classList.toggle("c3", false);

	// Polyfill for IE 10 and Firefox <24, where classList.toggle does not
	// support the second argument.
	if (testElement.classList.contains("c3")) {
		var _toggle = DOMTokenList.prototype.toggle;

		DOMTokenList.prototype.toggle = function(token, force) {
			if (1 in arguments && !this.contains(token) === !force) {
				return force;
			} else {
				return _toggle.call(this, token);
			}
		};

	}

	testElement = null;
}());

}

}

